### ISSUE TITLE
Support hierarchical softmax layer

### ISSUE THREAD

**Issue by thecortex** [OpenNMT/OpenNMT-py#541]
I'd like to customize the translation ML model by augmenting the RNN with a hierarchical softmax layer. I was wondering if a similar approach has been attempted before? If not is it better to inherit from model class and apply hierarchical softmax? Should I instead utilize the RNN hidden layers state instead? any guide for extending OpenNMT code without breaking it?

**Comment by helson73** [comment]
You can write a own Softmax class for doing that, and add an option to tell generator which version of softmax should be used.
I think you may need a pre-defined tree structure for vocab hierarchy.

**Comment by srush** [comment]
The sharding helps with memory, but not speed. In theory HSM should be faster, particularly on CPU. 

If you can find a good opensource HSM implementation in PyTorch, it should be as easy as just substituting it in for our Softmax Layer. Let us know if you find one.

**Comment by vince62s** [comment]
If you guys want to go this way, I strongly suggest "adaptive softmax", it is more efficient.
A few guys implemented it in pytorch. Just google "pytorch adaptive softmax".

**Comment by thecortex** [comment]
Thanks, @srush, @helson73, & @vince62s, I am still struggling to find out and I'm writing my own HSM to work with OpenNMT, but in the meantime, these are all I could find:

* Morin, F., & Bengio, Y. (2005, 2008). Hierarchical Probabilistic Neural Network Language Model.
* Distributed Representations of Words and Phrases and their Compositionality
Tomas Mikolov, Ilya Sutskever, Kai Chen, Greg Corrado, Jeffrey Dean (Submitted on 16 Oct 2013)

* https://github.com/lateral/hs-semantic-tree
* https://github.com/farizrahman4u/huffmax
* https://blog.lateral.io/2017/09/semantic-trees-hierarchical-softmax/

**Comment by thecortex** [comment]
I will be pushing my implementation over here and you could review as I go if you don't mind
https://github.com/thecortex/hierarchical-softmax-language-model

**Comment by thecortex** [comment]
Another resource suggests the following for efficincy and speed:
https://drive.google.com/file/d/0B7XkCwpI5KDYRWRnd1RzWXQ2TWc/edit

The full softmax can be replaced by:

1. Hierarchical softmax (Morin and Bengio)
2. Hinge loss (Collobert and Weston)
3. Noise contrastive estimation (Mnih et al.)
4. Negative sampling (our work)

**Comment by thecortex** [comment]
I have found this implementation which is much better documented and believe we can convert easily to PyTorch: https://code.google.com/archive/p/word2vec/source/default/source
https://code.google.com/archive/p/word2vec/

**Comment by thecortex** [comment]
[1] Tomas Mikolov, Kai Chen, Greg Corrado, and Jeffrey Dean. Efficient Estimation of Word Representations in Vector Space. In Proceedings of Workshop at ICLR, 2013.

[2] Tomas Mikolov, Ilya Sutskever, Kai Chen, Greg Corrado, and Jeffrey Dean. Distributed Representations of Words and Phrases and their Compositionality. In Proceedings of NIPS, 2013.

[3] Tomas Mikolov, Wen-tau Yih, and Geoffrey Zweig. Linguistic Regularities in Continuous Space Word Representations. In Proceedings of NAACL HLT, 2013.

**Comment by vince62s** [comment]
I certainly don't want to prevent you from working on HSM but honestly this is not really in the mood these days. There are better methods with Importance sampling or Adaptive softmax.

**Comment by thecortex** [comment]
Yes, you're right, I have to do it to compare against modern ways anyhow as part of my research thesis. I believe the hidden layer for RNN should have the same dimensionality as the input vector presentation

**Comment by srush** [comment]
Yeah, a lot of these are out of date. Maybe something like this to start.

https://github.com/rosinality/adaptive-softmax-pytorch

**Comment by thecortex** [comment]
How to shard the Huffman Tree as it will never fit in memory?

**Comment by thecortex** [comment]
How do we pass targets to forward to allow tree construction?

**Comment by vince62s** [comment]
closing this, but the best would be to use the now available in pytorch 0.4.1:
Adaptive Softmax: https://github.com/pytorch/pytorch/releases/tag/v0.4.1
